models/project.go

```diff
@@ -0,0 +1,289 @@
+}
+
+func getProjectByID(e Engine, id int64) (*Project, error) {
+	p := new(Project)
+
+	has, err := e.ID(id).Get(p)
+	if err != nil {
+		return nil, err
+	} else if !has {
+		return nil, ErrProjectNotExist{ID: id}
+	}
+
+	return p, nil
+}
+
+// UpdateProject updates project properties
+func UpdateProject(p *Project) error {
+	return updateProject(x, p)
+}
+
+func updateProject(e Engine, p *Project) error {
+	_, err := e.ID(p.ID).Cols(
+		"title",
+		"description",
+	).Update(p)
+	return err
+}
+
+func countRepoProjects(e Engine, repoID int64) (int64, error) {
+	return e.Where("repo_id=? AND type=?", repoID, ProjectTypeRepository).
+		Count(new(Project))
+}
+
+func countRepoClosedProjects(e Engine, repoID int64) (int64, error) {
+	return e.
+		Where("repo_id=? AND type=? AND is_closed=?", repoID, ProjectTypeRepository, true).
+		Count(new(Project))
+}
+
+// ChangeProjectStatus toggle a project between opened and closed
+func ChangeProjectStatus(p *Project, isClosed bool) error {
+
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	p.IsClosed = isClosed
+	p.ClosedDateUnix = timeutil.TimeStampNow()
+	if _, err := sess.ID(p.ID).Cols("is_closed", "closed_date_unix").Update(p); err != nil {
+		return err
+	}
+
+	numProjects, err := countRepoProjects(sess, p.RepoID)
+	if err != nil {
+		return err
+	}
+
+	numClosedProjects, err := countRepoClosedProjects(sess, p.RepoID)
+	if err != nil {
+		return err
+	}
+
+	if _, err = sess.ID(p.RepoID).Cols("num_projects, num_closed_projects").Update(&Repository{
+		NumProjects:       int(numProjects),
+		NumClosedProjects: int(numClosedProjects),
+	}); err != nil {
+		return err
+	}
+
+	return sess.Commit()
+}
+
+// DeleteProjectByID deletes a project from a repository.
+func DeleteProjectByID(id int64) error {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	if err := deleteProjectByID(sess, id); err != nil {
+		return err
+	}
+
+	return sess.Commit()
+}
+
+func deleteProjectByID(e Engine, id int64) error {
+	p, err := getProjectByID(e, id)
+	if err != nil {
+		if IsErrProjectNotExist(err) {
+			return nil
+		}
+		return err
+	}
+
+	if err := deleteProjectIssuesByProjectID(e, id); err != nil {
+		return err
+	}
+
+	if err := deleteProjectBoardByProjectID(e, id); err != nil {
+		return err
+	}
+
+	if _, err = e.ID(p.ID).Delete(new(Project)); err != nil {
+		return err
+	}
+
+	numProjects, err := countRepoProjects(e, p.RepoID)
+	if err != nil {
+		return err
+	}
+
+	numClosedProjects, err := countRepoClosedProjects(e, p.RepoID)
+	if err != nil {
+		return err
+	}
+
+	if _, err = e.ID(p.RepoID).Cols("num_projects, num_closed_projects").Update(&Repository{
+		NumProjects:       int(numProjects),
+		NumClosedProjects: int(numClosedProjects),
+	}); err != nil {
+		return err
+	}
+
+	if _, err = e.Exec("UPDATE `issue` SET project_id = 0 WHERE project_id = ?", p.ID); err != nil {
```

Comment:
Hmm... I think this may be out of date

Issue no longer has a project_id column. I guess this should be removed?